### DISCUSSION TITLE
Embedded UI

### DISCUSSION THREAD

**Discussion by taraktikos** [riverqueue/riverui#86] [General]
Hello, 
Is there a possibility to embed river ui as a handler to exiting project without deploying separate binary or docker image?

I mean something like this:

```
handlerOpts := &riverui.HandlerOpts{
    Client: riverClient,
    DBPool: dbRiverPool,
    Logger: slog.Default(),
    Prefix: "river-ui",
}

riverHandler, err := riverui.NewHandler(handlerOpts)
if err != nil {
    return fmt.Errorf("new river handler: %w", err)
}

e := echo.New()
e.GET("/river-ui", echo.WrapHandler(riverHandler))
```

I tried it but it doesn't work because of empty `dist` directory

**Comment by brandur** [comment]
Hi Taras, yes we're going to support this, but punted on it to get the initial release out the door.

As you've noticed, we need to include a TypeScript bundle for the whole setup to work. We're still working on a good way to do that, but we'll hopefully figure one out sooner rather than later.

**Reply by taraktikos** [reply]
Hi Brandur.
Thank you for the quick response. It sounds great!

**Reply by bgentry** [reply]
@taraktikos the main blocker is figuring how to get both the frontend React app in Vite and its Tanstack Router to properly support being mounted at arbitrary path prefixes. Right now they're statically configured at build time. I have made some attempts to fix this but each time I was unable to get it to work correctly and had to stop from going down the rabbithole.

If anybody out there is familiar enough with these aspects to get it working, it would be a hugely appreciated contribution :pray:

**Comment by rahulpr22** [comment]
Just checking in to see if there’s been any progress on supporting River UI as an embeddable HTTP handler — similar to what is mentioned earlier in this thread.

We have a very similar use case where we’d like to embed the River UI directly into our existing Go service (without running a separate binary or Docker container)

**Reply by bgentry** [reply]
Having trouble finding the exact issue to cross reference but yes, this has been possible since late last year: https://riverqueue.com/docs/river-ui#embedding-into-another-go-app

**Reply by rahulpr22** [reply]
@bgentry Thanks for confirming! 

I initially struggled to set up the embedded River UI because I was using grpc-gateway’s runtime ServeMux (github.com/grpc-ecosystem/grpc-gateway/v2/runtime). After applying this fix, I was able to get the UI working properly.

I initially tried:
```
mux.HandlePath("GET", "/riverui", func(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
    s.riverUiServer.ServeHTTP(w, r)
})
```
…but that didn't work for nested routes like /riverui/assets/..., etc.

What worked instead was:
```
mux.HandlePath("GET", "/riverui/{rest=**}", func(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
    s.riverUiServer.ServeHTTP(w, r)
})
```

By adding `{rest=**}`, the handler captures every sub-route under /riverui, ensuring that all UI assets are served properly. Sharing in case this helps others encountering the same problem.
